Userland/Libraries/LibCore/SharedCircularQueue.h

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright (c) 2022, kleines Filmröllchen <filmroellchen@serenityos.org>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/Assertions.h>
+#include <AK/Atomic.h>
+#include <AK/BuiltinWrappers.h>
+#include <AK/Error.h>
+#include <AK/Format.h>
+#include <AK/Function.h>
+#include <AK/NonnullRefPtr.h>
+#include <AK/Platform.h>
+#include <AK/RefCounted.h>
+#include <AK/RefPtr.h>
+#include <AK/String.h>
+#include <AK/Types.h>
+#include <AK/Variant.h>
+#include <AK/Weakable.h>
+#include <LibCore/AnonymousBuffer.h>
+#include <LibCore/System.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <sched.h>
+#include <sys/mman.h>
+
+namespace Core {
+
+// A circular lock-free queue (or a buffer) with a single producer,
+// residing in shared memory and designed to be accessible to multiple processes.
+// This implementation makes use of the fact that any producer-related code can be sure that
+// it's the only producer-related code that is running, which simplifies a bunch of the synchronization code.
+// The exclusivity and liveliness for critical sections in this class is proven to be correct
+// under the assumption of correct synchronization primitives, i.e. atomics.
+// In many circumstances, this is enough for cross-process queues.
+// This class is designed to be transferred over IPC and mmap()ed into multiple processes' memory.
+// It is a synthetic pointer to the actual shared memory, which is abstracted away from the user.
+// clang-format off
+template<typename T, size_t Size = 32>
```

Comment:
`Size` here seems like what we call `Capacity` in other containers? Should we rename that for clarity?

Reply:
No, because there's no size-capacity split as in resizeable containers like Vector. For this class size = capacity, and as with Array and FixedArray, we should call it size.

Reply:
Then you should rename your `capacity()` member to `size()`? :thinking:

Reply:
:facepalm: